magefile.go

```diff
@@ -0,0 +1,350 @@
+		dstPath := filepath.Join("./proto", s)
+		return copy(path, dstPath)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func ProtocVersionCheck() error {
+	cmd, err := protocOutCmd()
+	if err != nil {
+		return err
+	}
+	out, err := cmd("--version")
+	if err != nil {
+		return err
+	}
+	protocVersion := strings.TrimSpace(strings.Split(out, " ")[1])
+	if len(strings.Split(protocVersion, ".")) != 3 {
+		return fmt.Errorf("expected a semver but got %s", protocVersion)
+	}
+	if protocVersion < PROTOC_VERSION_MIN {
+		return fmt.Errorf("found protoc version %s but the minimum required is %s", protocVersion, PROTOC_VERSION_MIN)
+	}
+	return nil
+}
+
+func ProtocInstall() error {
+	if ok, err := exists("./protoc"); ok || err != nil {
+		return err
+	}
+
+	archOs, err := protocArchOs()
+	if err != nil {
+		return err
+	}
+	url := fmt.Sprintf(
+		"https://github.com/protocolbuffers/protobuf/releases/download/v%s/protoc-%s-%s.zip",
+		PROTOC_VERSION_DOWNLOAD, PROTOC_VERSION_DOWNLOAD, archOs,
+	)
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.ContentLength < 1000 {
+		contents, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("failed to download protoc: unexpected response '%s'", string(contents))
+	}
+
+	f, err := os.Create("protoc.zip")
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(f, resp.Body)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll("protoc.zip")
+
+	return unzip("protoc.zip", "./protoc")
+}
+
+func unzip(zipPath, dstPath string) error {
+	read, err := zip.OpenReader(zipPath)
+	if err != nil {
+		return err
+	}
+	defer read.Close()
+	for _, file := range read.File {
+		name := path.Join(dstPath, file.Name)
+		if file.Mode().IsDir() {
+			os.MkdirAll(path.Dir(name), os.ModeDir)
+			continue
+		}
+		open, err := file.Open()
+		if err != nil {
+			return err
+		}
+		defer open.Close()
+		os.MkdirAll(path.Dir(name), os.ModeDir)
+		create, err := os.Create(name)
+		if err != nil {
+			return err
+		}
+		defer create.Close()
+		_, err = io.Copy(create, open)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func copy(srcPath, dstPath string) error {
+	os.MkdirAll(filepath.Dir(dstPath), os.ModeDir)
+	src, err := os.Open(srcPath)
+	if err != nil {
+		return err
+	}
+	dst, err := os.Create(dstPath)
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(dst, src)
+	return err
+}
+
+func protocOutCmd() (func(args ...string) (string, error), error) {
+	var protocPath string
+	if runtime.GOOS == "windows" {
+		protocPath = "./protoc/bin/protoc.exe"
+	} else {
+		protocPath = "./protoc/bin/protoc"
+	}
+	ok, err := exists(protocPath)
+	if err != nil {
+		return nil, err
+	}
+	if ok {
+		return sh.OutCmd(protocPath), nil
+	}
+	return sh.OutCmd("protoc"), nil
+}
+
+func exists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
```

Comment:
can you replace this block with just `return false, err`, caller anyways need to check is `err != nil` and you return false in both cases.

Reply:
Done